### PR TITLE
Add LocalToAzureDataLakeStorageOperator

### PR DESCRIPTION
**Pull request by ephraimbuddy**
This PR adds LocalToAzureDataLakeStorageOperator. This operator will help to add system test for ADLSToGCSOperator.
Also, the upload and download methods of AzureDataLakeHook was improved to accommodate extra arguments since the documentation allows for more arguments.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).

### CODE REVIEW

**Comment by ephraimbuddy** [comment]
@mik-laj, I plan to work on the missing example dags. This Operator is needed to complete the example dag and system test for adls_to_gcs. It will upload a file to Data Lake which will then be transferred to GCS. Kindly review. Thanks

**Comment by mik-laj** [comment]
@michalslowikowski00 Can you look at it?

**Comment by ephraimbuddy** [comment]
@mik-laj, @michalslowikowski00, @turbaszek please check if this is necessary :)

File: airflow/providers/microsoft/azure/transfers/local_to_adls.py [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+        self.nthreads = nthreads
+        self.buffersize = buffersize
+        self.blocksize = blocksize
+        self.extra_upload_options = extra_upload_options
+        self.azure_data_lake_conn_id = azure_data_lake_conn_id
+
+    def execute(self, context):
+        if '**' in self.local_path:
+            raise AirflowException("Recursive glob patterns using `**` are not supported")
+        if not self.extra_upload_options:
```
Comment (michalslowikowski00) [inline airflow/providers/microsoft/azure/transfers/local_to_adls.py:91]: Do you need to pass empty dict in execute?

File: airflow/providers/microsoft/azure/transfers/local_to_adls.py [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+from airflow.utils.decorators import apply_defaults
+
+
+class AzureDataLakeStorageUploadOperator(BaseOperator):
+    """
+    Upload file(s) to Azure Data Lake
+
+    :param local_path: local path. Can be single file, directory (in which case,
+            upload recursively) or glob pattern. Recursive glob patterns using `**`
+            are not supported.
```
Comment (michalslowikowski00) [inline airflow/providers/microsoft/azure/transfers/local_to_adls.py]: We are trying to avoid full stops in documentation.
```suggestion
            are not supported
```

**Comment by michalslowikowski00** [comment]
Looks good to me. 
Personally I would love to see:
- example DAG (https://github.com/apache/airflow/tree/master/airflow/providers/microsoft/azure/example_dags)
- system test (https://github.com/apache/airflow/tree/master/tests/providers/microsoft/azure/operators)
- documentation (https://github.com/apache/airflow/tree/master/docs/howto/operator)

There are no docs about Azure and Microsoft operators at all. You could be a pioneer and add docs and make a good example for future contributors. :)

File: airflow/providers/microsoft/azure/transfers/local_to_adls.py (hunk 1 above)
Comment (ephraimbuddy) [reply airflow/providers/microsoft/azure/transfers/local_to_adls.py:84]: Hi, sorry, I don't realy get the question. If it's about the extra_upload_options, I think that empty dictionary is Ok instead of None or what do you advise?

Comment (ephraimbuddy) [reply airflow/providers/microsoft/azure/transfers/local_to_adls.py:91]: Got it now, thanks

**Comment by ephraimbuddy** [comment]
> Looks good to me.
> Personally I would love to see:
> 
> * example DAG (https://github.com/apache/airflow/tree/master/airflow/providers/microsoft/azure/example_dags)
> * system test (https://github.com/apache/airflow/tree/master/tests/providers/microsoft/azure/operators)
> * documentation (https://github.com/apache/airflow/tree/master/docs/howto/operator)
> 
> There are no docs about Azure and Microsoft operators at all. You could be a pioneer and add docs and make a good example for future contributors. :)
> 
> I am sorry for such late review. :(

Yeah, I plan to actively work on example dags on this provider and some others but I'm thinking of starting it up on a separate PR. Currently, there's no SystemTest class for this provider. Do you suggest I add everything on this PR?

**Comment by michalslowikowski00** [comment]
> > Looks good to me.
> > Personally I would love to see:
> > 
> > * example DAG (https://github.com/apache/airflow/tree/master/airflow/providers/microsoft/azure/example_dags)
> > * system test (https://github.com/apache/airflow/tree/master/tests/providers/microsoft/azure/operators)
> > * documentation (https://github.com/apache/airflow/tree/master/docs/howto/operator)
> > 
> > There are no docs about Azure and Microsoft operators at all. You could be a pioneer and add docs and make a good example for future contributors. :)
> > I am sorry for such late review. :(
> 
> Yeah, I plan to actively work on example dags on this provider and some others but I'm thinking of starting it up on a separate PR. Currently, there's no SystemTest class for this provider. Do you suggest I start SystemTests with this PR or Create another one and do example dags with tests in separate PRs?

IMHO SystemTest could be in separate PR but example dag should be in this PR.

**Comment by ephraimbuddy** [comment]
Alright. Thanks

**Comment by michalslowikowski00** [comment]
if you need help, let me know.

File: airflow/providers/microsoft/azure/transfers/local_to_adls.py (hunk 1 above)
Comment (ephraimbuddy) [reply airflow/providers/microsoft/azure/transfers/local_to_adls.py:91]: This empty dictionary is necessary so I could use keyword args for the upload_file method

**Comment by ephraimbuddy** [comment]
I have added the example dag. Will add a doc and system test to verify the example in a separate PR as discussed @michalslowikowski00

File: airflow/providers/microsoft/azure/transfers/local_to_adls.py (hunk 1 above)
Comment (michalslowikowski00) [reply airflow/providers/microsoft/azure/transfers/local_to_adls.py:91]: Great. I see that this param is needed.

**Comment by michalslowikowski00** [comment]
> I have added the example dag. Will add a doc and system test to verify the example in a separate PR as discussed @michalslowikowski00

Great. Looks good for me.
Now someone from the community has to accept it.

File: airflow/providers/microsoft/azure/hooks/azure_data_lake.py [hunk 3]
```diff
@@ -77,7 +77,14 @@ def check_for_file(self, file_path):
...
     def upload_file(
-        self, local_path, remote_path, nthreads=64, overwrite=True, buffersize=4194304, blocksize=4194304
+        self,
+        local_path,
+        remote_path,
+        nthreads=64,
+        overwrite=True,
+        buffersize=4194304,
+        blocksize=4194304,
+        **kwargs,
```
Comment (turbaszek) [inline airflow/providers/microsoft/azure/hooks/azure_data_lake.py:89]: Can we add type annotations, please? 🙏

File: airflow/providers/microsoft/azure/hooks/azure_data_lake.py [hunk 4]
```diff
@@ -113,10 +120,18 @@ def upload_file(
...
     def download_file(
-        self, local_path, remote_path, nthreads=64, overwrite=True, buffersize=4194304, blocksize=4194304
+        self,
+        local_path,
+        remote_path,
+        nthreads=64,
+        overwrite=True,
+        buffersize=4194304,
+        blocksize=4194304,
+        **kwargs,
```
Comment (turbaszek) [inline airflow/providers/microsoft/azure/hooks/azure_data_lake.py:136]: Same here, otherwise those changes are not related

File: airflow/providers/microsoft/azure/transfers/local_to_adls.py [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+from typing import Dict, Any, Optional
+from airflow.exceptions import AirflowException
+from airflow.models import BaseOperator
+from airflow.providers.microsoft.azure.hooks.azure_data_lake import AzureDataLakeHook
+from airflow.utils.decorators import apply_defaults
+
+
+class LocalToAzureDataLakeStorageOperator(BaseOperator):
+    """
+    Upload file(s) to Azure Data Lake
```
Comment (turbaszek) [inline airflow/providers/microsoft/azure/transfers/local_to_adls.py:27]: Can you add a reference to how to guide?

File: airflow/providers/microsoft/azure/transfers/local_to_adls.py [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+    :param azure_data_lake_conn_id: Reference to the Azure Data Lake connection
+    :type azure_data_lake_conn_id: str
+    """
+
+    @apply_defaults
+    def __init__(
+        self,
+        *,
+        local_path: str,
+        remote_path: str,
```
Comment (turbaszek) [inline airflow/providers/microsoft/azure/transfers/local_to_adls.py:69]: I would suggest adding those fields to template fields.

**Comment by ephraimbuddy** [comment]
Hi @turbaszek, please can you take a look once more, thanks.

**Review by turbaszek — APPROVED** [review]
